docs/phdata.rst

```diff
@@ -483,26 +555,92 @@ Also, if some information is not available the relative field may be omitted.
 - **software**: (string) Software used to save the original data file.
 - **software_version**: (string) Version of the software used to save the original data file.
 
+.. _user_group:
+
+User group
+----------
+
+The ``/user/`` group is reserved for user defined fields.
+There are no specific definitions for the groups contained within `user`.
+Nor will standard readers of *photon-HDF5* files use any fields withing this group.
+
+This group is intended to be used to store data not covered by the official photon-HDF5
+specification. While there is not required format for this group, some suggestions are
+provided.
+
+If the data is being converted from a format like .ptu from picoquant or .spc from
+Beckr&Hickl, some metadata stored in the file is not included in the official photon-HDF5
+specification. This metadata should be stored in a group named by the company manufacturing
+the original data file (i.e. `picoquant` for .ptu files, `becker_hickl` for .spc files).
+
+The ``user`` group can also be used to store results of various type of analysis.
+Such data should be stored in a sub-group named according to the software used
+for such analysis.
+For instance, FRETBursts, if directed to do so, will save results of burst
+analysis under the group ``/user/FRETBursts``.
+
+.. _exp_settings:
+
+Experimental Settings User Group
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+It is often desirable to store additional information and/or explanation of the experiment.
+Such information is best stored in a ``user/experimental_settings`` group. All information in this
+group should be stored in a way that is self-explanatory, with human-readable details included.
+
+If :ref:`marker IDs<marker_ids>`) are included in ``/photon_data/detectors``,
+``/user/experimental_settings/`` is the preferred location to specify the significance
+of each marker ID.
+
+If values such as FRET correction factors are stored, the specific definition and method by which
+they were determined should be included as well so that interpretation can be made unambiguous.
+We suggest using the definition in `Hellenkamp et. al. 2018 <https://doi.org/10.1038/s41592-018-0085-0>`_
+
+See <link to notebook to be created> for an example of how to use the */user/experimental_settings/**
+group.
+
 
 .. _glossary:
 
 Additional notes and definitions
 --------------------------------
 
-.. _detector_ids:
+.. _record_ids:
 
-Detector pixel IDs
-^^^^^^^^^^^^^^^^^^
+Record IDs
+^^^^^^^^^^
```

Comment:
The HDF5 field names already use "detector" for all of these IDs (the `/photon_data/detectors` array contains both photon and non-photon), so I wonder if it will be less confusing if the general name just remains "detector ID"; we could use "photon detector ID" and "non-photon detector ID" when only one kind is meant.

I understand it's a tradeoff between using terms that better describe the thing vs terms that better match the HDF5 fields.

(I'm also slightly uncomfortable with using "record" for this, because that word is often used for a single TTTR record (often 32 bits) -- a photon, marker(s), or counter overflow.)

Reply:
You are discussing adding "markers/non-detector" flags and are wondering how to shove them into a array that was explicitly designed to store information on which detector a timestamp is coming from.
No surprise finding the proper terminology is confusing. :-)
When encountering such a problem, try to think out of the box.
My radical suggestion is to postpone introducing this to until after we have understood each other and weighted the pros and cons of adding this level of complexity (i.e. until 0.6).
But as an alternative to the above suggestion of trying to insert information of a different nature in an existing structure, I'll bring up the possibility to add an optional array of "markers" to the photon_data group(s).
If there is no markers (current files), there is no array.
If there are markers (future files), there is an array. It will mostly be filled with '0' (no marker), and those entry that are non '0' will be associated with a dummy "detector" entry (i.e. any value works as it will be ignored, but one could use the convention that 2^16-1 will be the equivalent of an integer NaN), since there is a priori no reason to have any specific detector associated with such a marker.
If I am not mistaken, HDF5 should be able to compress sparse "marker" arrays very efficiently.

So to answer your question: if there is no marker introduction in this version (0.5), then the question becomes moot.

Reply:
I don't see a huge problem with "detector" not necessarily meaning "photon detector" (hence my suggestion above). It can be an (electrical pulse) edge detector or whatever. All we need is to explicitly define the term "detector" to be more general (as @harripd has already done under different terminology). IMHO "detector" is actually a pretty good term to mean "something that produces timestamps (_records_) for observed _events_ that are otherwise indistinguishable".

My concern, to reiterate, was with letting the terminology diverge between the HDF5 names and the spec, which makes it harder to keep the definitions straight in one's head.

(Using code-font for HDF5 fields and quotes for spec terms) if the thing is called `detector` in the HDF5, you need to think in your head that `detector` can be either photon or non-photon anyway. I fear that adding to this situation a different name "record" for the same concept as `detector`, and especially the same name "detector (ID)" for a different concept (a `detector` that is not a "marker"), is going to increase the risk of people misinterpreting the spec.

---

I don't think using a separate `markers` array is going to win us much, I'm afraid. It is overall more complicated, it still requires the `detectors` array to contain a value that photon-only readers need to skip, and (at least as proposed) it imposes the irregular requirement that ID 0 cannot be a marker, which will require special-casing in all handling code (or bugs if not done right).

---

Regarding using "event" instead of "record" -- I think I have the same problem as with "record": both words suggest to me a single (photon or marker) event rather than a channel of input data.

("Event" and "record" would both be good words for talking about a single entry across the `timestamps` and `detectors` (and `nanotimes`) arrays, when not distinguishing between photons and non-photons. The term "record" could also be applied to other places where we have parallel arrays of the same length that act as the _columns_ of a table (such as `/setup/detectors`): a record would denote a _row_ in the conceptual table. That's clearly a concept in this spec but we don't have a word for it currently. I'd suggest saving these words for improving the readability of the spec in the future.)

Reply:
Regarding a separate array: being sparse and for instance using 0 as the "no marker" value, it makes it extremely easy and fast to slice the actual photon_data array (and its associated detectors array, if any).
The MAIN advantage I see in it, is that it does not confuse a reader that is not aware of those additional markers (as I mentioned, since the "detector" of a "marker" is irrelevant, it can be anything : no need for a special value as I was musing about).
Yes, you will add "dummy" photons in your processed data, but  it probably won't matter if there are very few of them.
This is kind of the essence of a backward compatible format.
You can ignore what you don't know about and be fine.
With a "detectors" array containing values that are not detector IDs, but pass as such (for an unsuspecting code), a software trying to interpret this will hallucinate detectors which will eventually contain very few time stamps.